### PR TITLE
Explain code in flow.cylc files.

### PR DESCRIPTION
**Pull request by hjoliver**
Explain exactly what Jinja2 and shell code embedded in a `flow.cylc` means, and when/where it will be evaluated.

**Requirements check-list**
- [x] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.

### CODE REVIEW

File: src/user-guide/writing-workflows/configuration.rst [hunk 1]
```diff
@@ -62,10 +62,137 @@ documentation, configuration files, etc. When the workflow is :ref:`installed
...
+
+Cylc workflow configurations are not executable scripts or programs. Rather,
+they configure, at start-up, the Cylc scheduler program to run a particular
+workflow.
+
+The scheduler understands a static syntax that contains only ``[section
+headings]`` and configuration items of the form ``platform = hpc1``. These
+configuration items are not even "variables" that can be referenced elsewhere
+in the file. Some values may look complicated, but they are still just text
+strings with content that is meaningful to the scheduler.
```
Comment (oliver-sanders) [inline src/user-guide/writing-workflows/configuration.rst:78]: The tone here is a bit negative, Cylc provides a powerful configuration format, it just isn't dynamic.

Maybe just explain that the Cylc config is static, but can be augmented by template engines such as Jinja2 and explain that these engines are run *before* startup.

File: src/user-guide/writing-workflows/configuration.rst [hunk 2]
```diff
@@ -62,10 +62,137 @@ documentation, configuration files, etc. When the workflow is :ref:`installed
...
+
+Nevertheless a `flow.cylc` file may appear to contain several kinds of embedded
+code:
+
+    - Python-like Jinja2 or EmPy code, such as ``{% set PLANET = "earth" %}``
+    - Bash shell variable assignments such as ``PLANET=${PLANET:-earth}``
+    - Bash shell scripting, such as ``script = "run-model.exe /path/to/data"``
```
Comment (oliver-sanders) [inline src/user-guide/writing-workflows/configuration.rst:85]:
```suggestion
- Python-like Jinja2 or EmPy code, such as ``{% set PLANET = "earth" %}``
- Bash shell variable assignments such as ``PLANET=${PLANET:-earth}``
- Bash shell scripting, such as ``script = "run-model.exe /path/to/data"``
```

File: src/user-guide/writing-workflows/configuration.rst [hunk 3]
```diff
@@ -62,10 +62,137 @@ documentation, configuration files, etc. When the workflow is :ref:`installed
...
+
+The documentation below explains exactly what these mean in the context of a
+workflow configuration, and when the code is evaluated.
+
+
+Jinja2 (or EmPy) Code
+^^^^^^^^^^^^^^^^^^^^^
```
Comment (oliver-sanders) [inline src/user-guide/writing-workflows/configuration.rst:92]: Link to the Jinja2/Empty bits of the user guilde from somewhere in here.

File: src/user-guide/writing-workflows/configuration.rst [hunk 4]
```diff
@@ -62,10 +62,137 @@ documentation, configuration files, etc. When the workflow is :ref:`installed
...
+programmatically generating the static configuration format required by the
+scheduler.
+
+The template processor manipulates the surrounding text (which to it is
+entirely arbitrary) and the result must be valid configuration syntax that
+will pass validation by Cylc.
+
+This is a preprocessing step. Jinja2 code does not execute as the workflow runs.
+
+Use ``cylc view -j`` to see the result, with all Jinja2 code "processed out".
```
Comment (oliver-sanders) [inline src/user-guide/writing-workflows/configuration.rst:105]: I think the `--process` option is the more generic advice?

```suggestion
Use ``cylc view -p`` to see the result, with all Jinja2 code "processed out".
```

File: src/user-guide/writing-workflows/configuration.rst [hunk 5]
```diff
@@ -62,10 +62,137 @@ documentation, configuration files, etc. When the workflow is :ref:`installed
...
+
+.. code-block:: cylc
+
+   # flow.cylc
+   {% set PLATFORM = "hpc1" %}
+   platform = {{ PLATFORM }}  # OK
+   {{ "Platform is: " ~ PLATFORM }} # OK
+ 
+Jinja2 code does not have access to the environment via normal shell syntax
+either, but you can read the local environment with native Jinja2 syntax:
```
Comment (oliver-sanders) [inline src/user-guide/writing-workflows/configuration.rst:127]: This is duplicating the Jinja2 section somewhat.

More positive spin:

```suggestion
The scheduler's environment is available via the ``environ`` dictionary.
```

**Review by oliver-sanders — COMMENTED** [review]
I fear the tone and length of this article makes Cylc seem limited and complex, which I don't think is right.

The main takeaway points are:

* Template processing happens at workflow startup.
* So you can't use Jinja2 to dynamically generate job scripts.
* Template engines don't have access to the worflow config, but you can provide information to it using variables.
* `cylc view` is a useful tool when working with template engines.

I would be tempted to convey these messages through a couple of notes or warnings in the Jinja2 section / tutorial.

**Comment by hjoliver** [comment]
>  I fear the tone and length of this article makes Cylc seem limited and complex, which I don't think is right.

On re-reading it, I agree.  I'll try to simplify and re-tone it!

> The main takeaway points are: ... I would be tempted to convey these messages through a couple of notes or warnings in the Jinja2 section / tutorial.

It's worse than that though, and involves shell code as well as Jinja2. 

Take this recent forum post from an (experienced!) user:

https://cylc.discourse.group/t/how-do-you-use-a-variable-defined-in-environment-in-the-directives-section/760

and shortly after:

https://cylc.discourse.group/t/how-do-you-use-a-variable-defined-in-environment-in-the-directives-section/760/3

**Comment by hjoliver** [comment]
OK thanks for the feedback. I've stripped that sucker back to the bare bones (I think). Try it now for size.

File: src/user-guide/writing-workflows/configuration.rst [hunk 6]
```diff
@@ -62,10 +62,53 @@ documentation, configuration files, etc. When the workflow is :ref:`installed
...
+Understanding Code in Workflow Configurations
+---------------------------------------------
+
+A workflow configuration is not executable code. It configures the scheduler
+program to run your workflow. But a `flow.cylc` file may contain:
+
+    - Embedded Python-like Jinja2 or EmPy templating code, such as
+      ``{% set PLANET = "earth" %}``
+    - Bash shell variable assignments and scripting, such as
+      ``script = "run-model.exe /path/to/data"``
```
Comment (oliver-sanders) [inline src/user-guide/writing-workflows/configuration.rst:76]:
```suggestion
program to run your workflow. A `flow.cylc` file may contain:

- Embedded Python-like Jinja2 or EmPy templating code, such as
  ``{% set PLANET = "earth" %}``
- Bash shell variable assignments and scripting, such as
  ``script = "run-model.exe /path/to/data"``
```

File: src/user-guide/writing-workflows/configuration.rst [hunk 7]
```diff
@@ -62,10 +62,53 @@ documentation, configuration files, etc. When the workflow is :ref:`installed
...
+
+    - Jinja2 code can only reference Jinja2 variables, not Cylc config items.
+    - Jinja2 (like Python) has its own syntax for reading environment variables.
+    - Jinja2 code that reads the environment or the filesystem will do so
+      during config file parsing on the scheduler run host, not on job hosts.
+      Beware of doing this in task definitions - do you want the scheduler
+      environment to affect shell code that runs in the job environment?
+    - Shell code destined for the job script can read the environment or
+      access the filesystem as the job runs on the job host, not on the
+      scheduler host.
```
Comment (oliver-sanders) [inline src/user-guide/writing-workflows/configuration.rst:105]:
```suggestion
- Jinja2 code is evaluated once when the workflow is started.
- Jinja2 code can only reference Jinja2 variables, not Cylc config items.
- Jinja2 (like Python) has its own syntax for reading environment variables.
- Jinja2 code that reads the environment or the filesystem will do so
  during config file parsing on the scheduler run host, not on job hosts.
  Beware of doing this in task definitions - do you want the scheduler
  environment to affect shell code that runs in the job environment?
- Shell code destined for the job script can read the environment or
  access the filesystem as the job runs on the job host, not on the
  scheduler host.
```

**Review by oliver-sanders — APPROVED** [review]
Cheers

**Comment by hjoliver** [comment]
TODO: add cross-referencing.

**Pull request merged**